base/subarray.jl

```diff
@@ -58,29 +61,39 @@ parent(a::AbstractArray) = a
 parentindexes(a::AbstractArray) = ntuple(i->OneTo(size(a,i)), ndims(a))
 
 ## SubArray creation
-# Drops singleton dimensions (those indexed with a scalar)
-function view{T,N}(A::AbstractArray{T,N}, I::Vararg{ViewIndex,N})
+# We always assume that the dimensionality of the parent matches the number of
+# indices that end up getting passed to it, so we store the parent as a
+# ReshapedArray view if necessary. The trouble is that `CartesianIndex`s make
+# the computation of the number of effective indices non-trivial. At this point
+# in the bootstrap, we don't have the required tools to do the computation,
+# so we just define `view` for non-CartesianIndex types here. The hard part is
+# done in multidimensional.jl after CartesianIndex gets defined properly.
+_maybe_reshape_parent(A::AbstractArray, ::NTuple{1, Bool}) = reshape(A, Val{1})
+_maybe_reshape_parent{_,N}(A::AbstractArray{_,N}, ::NTuple{N, Bool}) = A
+_maybe_reshape_parent{N}(A::AbstractArray, ::NTuple{N, Bool}) = reshape(A, Val{N}) # TODO: DEPRECATE FOR #14770
+function view(A::AbstractArray, I::ViewIndex...)
     @_inline_meta
     @boundscheck checkbounds(A, I...)
-    unsafe_view(A, I...)
-end
-function view(A::AbstractArray, i::ViewIndex)
-    @_inline_meta
-    @boundscheck checkbounds(A, i)
-    unsafe_view(reshape(A, Val{1}), i)
-end
-function view{N}(A::AbstractArray, I::Vararg{ViewIndex,N}) # TODO: DEPRECATE FOR #14770
-    @_inline_meta
-    @boundscheck checkbounds(A, I...)
-    unsafe_view(reshape(A, Val{N}), I...)
+    unsafe_view(_maybe_reshape_parent(A, index_ndims(I...)), I...)
 end
 
-function unsafe_view{T,N}(A::AbstractArray{T,N}, I::Vararg{ViewIndex,N})
+function unsafe_view(A::AbstractArray, I::ViewIndex...)
     @_inline_meta
     J = to_indexes(I...)
     SubArray(A, J, map(unsafe_length, index_shape(A, J...)))
 end
-function unsafe_view{T,N}(V::SubArray{T,N}, I::Vararg{ViewIndex,N})
+# Reindexing by CartesianIndex is hard, so flatten them out first
+unsafe_view(V::SubArray, I::ViewIndex...) = unsafe_view(V, IteratorsMD.flatten(I)...)
```

Comment:
This type of construct is a little bit vulnerable to `StackOverflowErrors`, particularly if someone later changes the definitions of `ViewIndex` above. What about splitting types this way:
```jl
typealias IndexWeCanHandleHere ...

unsafe_view(V::SubArray, I::Union{IndexWeCanHandleHere,AbstractCartesianIndex}...) = unsafe_view(V, IteratorsMD.flatten(I)...)
unsafe_view(V::SubArray, I::IndexWeCanHandleHere...) = # the "real" implementation
```
That way, since you know that `flatten` will take care of `AbstractCartesianIndex`, you've future-proofed this definition.